### ISSUE TITLE
dropConnection appears to leave connections around

### ISSUE THREAD

**Issue by meejah** [crossbario/autobahn-python#1064]
Found by ashleyk_ and reported in IRC.

In `failHandshake` in `autobahn/websocket/protocol.py` we call `self.dropConnection(abort=False)`. This then ultimately calls `transport.loseConnection()`. If using TLS, these connections don't actually close (in Twisted at least).

An easy fix is to pass `abort=True` at `protocol.py:3020`. Could even be a Twisted issue..`ashleyk_` had made an example: https://paste.debian.net/plain/1047301

**Comment by glyph** [comment]
Perhaps Twisted has registered a producer on something at that point in the handshake?

**Comment by meejah** [comment]
So far, in debugging / repeating what `ashleyk_` reported, it only happens on TLS and there is one producer registered (on the `HTTPChannel`). I haven't yet shortened this to a minimal example, though

**Comment by glyph** [comment]
If you register a producer and then call `loseConnection`, the connection is not actually closed until the producer unregisters itself.

This is working as designed.  The idea is that a producer "contains" some amount of bytes, and in the same way that `transport.write(b'data'); transport.loseConnection()` will deliver `b'data'` to the underlying transport before closing it, even though it "should be" closed, `transport.registerProducer(ProducerContainingStaticData(b'data')); transport.loseConnection()` will also expect `ProducerContainingStaticData` to illustrate that it is "done" by unregistering itself, so that we can ensure `b'data'` is written, even if there's only buffer space for `b'da`` at the moment of registration.

**Comment by meejah** [comment]
So, if I add a call to `self.unregisterProducer()` here: https://github.com/crossbario/autobahn-python/blob/master/autobahn/websocket/protocol.py#L841 then the connections are closed/dropped immediately, as expected.

However, I'm not convinced if this is the right approach -- "we" didn't add any producers, so perhaps there's some better way to do this? (We are, however, doing some funky stuff to "take over" a transport: https://github.com/crossbario/autobahn-python/blob/master/autobahn/twisted/resource.py#L113)

Of note, the above "funky stuff" does not appear to work properly for HTTP2 -- I'll file a separate ticket for that. (`request.transport` is `None` in the HTTP/2 case).

**Comment by dimitar-kunchev** [comment]
@meejah You are right - that makes a difference. Added a call to self.unregisterProducer() at the end of the dropConnection method. Testing with the echo example process memory is fixed and doesn't go up. In fact I tried again with printing `__del__` calls and occasionally I see the gc reaping them as expected.
Note that testing with `__del__` on python 3.5 should be safe ([PEP 442](https://www.python.org/dev/peps/pep-0442/))
